modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,332 @@
+// update set may be only partially written. Partially written update sets are
+// simply ignored when reading the journal.
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/persist"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+var journalMeta = persist.Metadata{
+	Header:  "Contractor Journal",
+	Version: "1.0.0",
+}
+
+// A journal is a log of updates to a JSON object.
+type journal struct {
+	f        *os.File
+	filename string
+}
+
+// update applies the updateSet atomically to j. It syncs the underlying file
+// before returning.
+func (j *journal) update(us updateSet) error {
+	if err := json.NewEncoder(j.f).Encode(us); err != nil {
+		return err
+	}
+	return j.f.Sync()
+}
+
+// Checkpoint refreshes the journal with a new initial object. It syncs the
+// underlying file before returning.
+func (j *journal) checkpoint(data contractorPersist) error {
+	// write to a new temp file
+	tmp, err := os.Create(j.filename + "_tmp")
+	if err != nil {
+		return err
+	}
+	enc := json.NewEncoder(tmp)
+	if err := enc.Encode(journalMeta); err != nil {
+		return err
+	}
+	if err := enc.Encode(data); err != nil {
+		return err
+	}
+	if err := tmp.Sync(); err != nil {
+		return err
+	}
+
+	// atomically replace the old file with the new one
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+	if err := j.f.Close(); err != nil {
+		return err
+	}
+	if err := os.Rename(tmp.Name(), j.filename); err != nil {
+		return err
+	}
+
+	// reopen the journal
+	j.f, err = os.OpenFile(j.filename, os.O_RDWR|os.O_APPEND, 0)
+	return err
+}
+
+// Close closes the underlying file.
+func (j *journal) Close() error {
+	return j.f.Close()
+}
+
+// newJournal creates a new journal, using an empty contractorPersist as the
+// initial object.
+func newJournal(filename string) (*journal, error) {
+	f, err := os.Create(filename)
+	if err != nil {
+		return nil, err
+	}
+	// write metadata
+	enc := json.NewEncoder(f)
+	if err := enc.Encode(journalMeta); err != nil {
+		return nil, err
+	}
+	// write empty contractorPersist
+	if err := enc.Encode(contractorPersist{}); err != nil {
+		return nil, err
+	}
+	if err := f.Sync(); err != nil {
+		return nil, err
+	}
+	return &journal{f: f, filename: filename}, nil
+}
+
+// openJournal opens the supplied journal and decodes the reconstructed
+// contractorPersist into data.
+func openJournal(filename string, data *contractorPersist) (*journal, error) {
+	// open file handle for reading and writing
+	f, err := os.OpenFile(filename, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// decode metadata
+	dec := json.NewDecoder(f)
+	var meta persist.Metadata
+	if err = dec.Decode(&meta); err != nil {
+		return nil, err
+	} else if meta.Version != journalMeta.Version {
+		return nil, errors.New("incompatible version")
```

Comment:
the persist package currently returns `ErrBadVersion` and `ErrBadHeader`, which are equally unhelpful. They do allow you to check for a specific error though. That may be more useful when we implement a more robust compatibility pipeline